From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Qing Ju <juqing@amazon.com>
Date: Wed, 7 Apr 2021 17:50:46 -0700
Subject: [PATCH] --EKS-PATCH-- Exposed etcd client auto-sync as kube-apiserver
 option

--EKS-PRIVATE--

Expose auto-sync option so that people can enable/disable auto-sync and
change auto-sync interval by setting flag --etcd-auto-sync-interval
in pod kube-apiserver.

cr: https://code.amazon.com/reviews/CR-51146892
---
 staging/src/k8s.io/apiserver/pkg/server/options/etcd.go        | 3 +++
 .../src/k8s.io/apiserver/pkg/storage/storagebackend/config.go  | 3 ++-
 .../apiserver/pkg/storage/storagebackend/factory/etcd3.go      | 1 +
 3 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/staging/src/k8s.io/apiserver/pkg/server/options/etcd.go b/staging/src/k8s.io/apiserver/pkg/server/options/etcd.go
index d8b45b8198f..2c1c27e1592 100644
--- a/staging/src/k8s.io/apiserver/pkg/server/options/etcd.go
+++ b/staging/src/k8s.io/apiserver/pkg/server/options/etcd.go
@@ -155,6 +155,9 @@ func (s *EtcdOptions) AddFlags(fs *pflag.FlagSet) {
 	fs.StringSliceVar(&s.StorageConfig.Transport.ServerList, "etcd-servers", s.StorageConfig.Transport.ServerList,
 		"List of etcd servers to connect with (scheme://ip:port), comma separated.")
 
+	fs.DurationVar(&s.StorageConfig.Transport.AutoSyncInterval, "etcd-auto-sync-interval", s.StorageConfig.Transport.AutoSyncInterval,
+		"The interval at which to update etcd endpoints to the latest members as reported by the cluster. Disabled when set to 0. Disabled by default.")
+
 	fs.StringVar(&s.StorageConfig.Prefix, "etcd-prefix", s.StorageConfig.Prefix,
 		"The prefix to prepend to all resource paths in etcd.")
 
diff --git a/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go b/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go
index 500e55c0206..1036be45404 100644
--- a/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go
@@ -37,7 +37,8 @@ const (
 // TransportConfig holds all connection related info,  i.e. equal TransportConfig means equal servers we talk to.
 type TransportConfig struct {
 	// ServerList is the list of storage servers to connect with.
-	ServerList []string
+	ServerList       []string
+	AutoSyncInterval time.Duration
 	// TLS credentials
 	KeyFile       string
 	CertFile      string
diff --git a/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go b/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go
index 9b1b5ee9dd9..83aa620ddce 100644
--- a/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go
@@ -151,6 +151,7 @@ func newETCD3Client(c storagebackend.TransportConfig) (*clientv3.Client, error)
 		dialOptions = append(dialOptions, grpc.WithContextDialer(dialer))
 	}
 	cfg := clientv3.Config{
+		AutoSyncInterval:     c.AutoSyncInterval,
 		DialTimeout:          dialTimeout,
 		DialKeepAliveTime:    keepaliveTime,
 		DialKeepAliveTimeout: keepaliveTimeout,
